### PR TITLE
Filer - Add firmware upgrade API

### PR DESCRIPTION
**Pull request by ygalblum**
Add UT

### CODE REVIEW

File: cterasdk/edge/firmware.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+
+class Firmware(BaseCommand):
+    """ Gateway Firmware upgrade API """
+
+    def __init__(self, gateway):
+        super().__init__(gateway)
+        self._filesystem = FileSystem.instance()
+
+    def upgrade(self, file_path, reboot=True, wait_for_reboot=True):
```
Comment (saimonation) [inline cterasdk/edge/firmware.py:19]: When uploading a firmware directly to the edge filer, it automatically reboots. I'm not sure the ``reboot=True`` flag is necessary.

File: cterasdk/edge/firmware.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        with open(file_path, 'rb') as fd:
+            return self._gateway.upload(
+                'proc/firmware',
+                dict(
+                    name='upload',
+                    firmware=(file_info['name'], fd, file_info['mimetype'][0])
+                )
+            )
+
+    def _wait_for_completion(self, task_pointer):
```
Comment (saimonation) [inline cterasdk/edge/firmware.py:45]: There's an edge/taskmgr.py module used to monitor the progress of background tasks. Can you use that instead?

File: cterasdk/edge/firmware.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+class Firmware(BaseCommand):
+    """ Gateway Firmware upgrade API """
+
+    def __init__(self, gateway):
+        super().__init__(gateway)
+        self._filesystem = FileSystem.instance()
+
+    def upgrade(self, file_path, reboot=True, wait_for_reboot=True):
+        """
+        Upgrade the Filer F/W with the provided file
```
Comment (saimonation) [inline cterasdk/edge/firmware.py:21]: I'd change F/W to firmware

File: cterasdk/edge/firmware.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+        :param str file_path: Path to the local file to upload
+        :param bool,optional reboot: Perform reboot after uploading the new F/W, defaults to True
```
Comment (saimonation) [inline cterasdk/edge/firmware.py:24]: Here too

File: cterasdk/edge/firmware.py (hunk 1 above)
Comment (ygalblum) [reply cterasdk/edge/firmware.py:19]: The UI calls the reboot API. If you don't call it, then the new F/W just sits there waiting for the next reboot

File: cterasdk/edge/firmware.py (hunk 2 above)
Comment (ygalblum) [reply cterasdk/edge/firmware.py:45]: I wanted to. But, the mechanism is handled differently -
1. The object returned from the upload command is different
2. Plus, it contains the entire URL (not a bgtask with UID)
3. The object returned from the Get status URL is different
4. All of this is specific to the firmware upgrade sequence (I looked at the G/W code)
As a result, this required different code that is not generic and hence will not be used anywhere else. So, I don't see the advantage of putting it anywhere outside of the scope of this class

File: cterasdk/edge/firmware.py (hunk 3 above)
Comment (ygalblum) [reply cterasdk/edge/firmware.py]: Done

File: cterasdk/edge/firmware.py (hunk 4 above)
Comment (ygalblum) [reply cterasdk/edge/firmware.py]: Done

File: cterasdk/edge/firmware.py (hunk 2 above)
Comment (saimonation) [reply cterasdk/edge/firmware.py:45]: I see. Ok. Btw, there's also a ``edge/tracker.py`` for scenario's where you monitor a specific schema attribute (not a BgTask). But perhaps it's not applicable in this scenario either.

**Pull request merged**